File: modules/ballerina-core/src/main/java/org/wso2/ballerina/core/debug/DebugManager.java
/*
*  Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
*
*  WSO2 Inc. licenses this file to you under the Apache License,
*  Version 2.0 (the "License"); you may not use this file except
*  in compliance with the License.
*  You may obtain a copy of the License at
*
*    http://www.apache.org/licenses/LICENSE-2.0
*
*  Unless required by applicable law or agreed to in writing,
*  software distributed under the License is distributed on an
*  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
*  KIND, either express or implied.  See the License for the
*  specific language governing permissions and limitations
*  under the License.
*/

package org.wso2.ballerina.core.debug;

import com.google.gson.Gson;
import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import io.netty.channel.Channel;
import org.wso2.ballerina.core.debug.dto.CommandDTO;
import org.wso2.ballerina.core.debug.dto.MessageDTO;
import org.wso2.ballerina.core.interpreter.nonblocking.debugger.BLangExecutionDebugger;
import org.wso2.ballerina.core.interpreter.nonblocking.debugger.BreakPointInfo;
import org.wso2.ballerina.core.model.NodeLocation;


import java.util.ArrayList;
import java.util.concurrent.Semaphore;

import static java.lang.Thread.sleep;

/**
 * {@code DebugManager} Manages debug sessions and handle debug related actions
 *
 * @since 0.8.0
 */
public class DebugManager {
    /**
     * The Execution sem. used to block debugger till client connects.
     */
    public volatile Semaphore executionSem;

    private DebugServer debugServer;

    /**
     * Object to hold debug session related context
     * @todo , allow managing multiple debug sessions.
     */
    private DebugSession debugSession;

    private static DebugManager debugManagerInstance = null;

    private boolean waitingForClient;

    /**
     * Instantiates a new Debug manager.
     */
    protected DebugManager() {
        executionSem = new Semaphore(0);
    }

    /**
     * Debug manager singleton
     *
     * @return DebugManager instance
     */
    public static DebugManager getInstance() {
        synchronized (DebugManager.class) {
            if (debugManagerInstance == null) {
                debugManagerInstance = new DebugManager();
            }
        }
        return debugManagerInstance;
    }

    /**
     * Initializes the debug manager single instance.
     */
    public void init() {
        // start the debug server if it is not started yet.
        if (this.debugServer == null) {
            this.debugServer = new DebugServer();
            this.debugServer.startServer();
        }
    }

    /**
     *  Wait till client connection establish.
     */
    public void waitTillClientConnect() {
        this.waitingForClient = true;
        //suspend the current thread till client connects.
        try {
            executionSem.acquire();
        } catch (InterruptedException e) {
            //todo proper error handling
        }
    }

    /**
     * Process debug command.
     *
     * @param json the json
     */
    public void processDebugCommand(String json) {
        BLangExecutionDebugger debugger = debugSession.getDebugger();
        Gson gson = new Gson();
        CommandDTO command = gson.fromJson(json, CommandDTO.class);
        switch (command.getCommand()) {
            case DebugConstants.CMD_RESUME:
                debugger.resume();
                break;
            case DebugConstants.CMD_STEP_OVER:
                debugger.stepOver();
                break;
            case DebugConstants.CMD_STEP_IN:
                debugger.stepIn();
                break;
            case DebugConstants.CMD_STEP_OUT:
                debugger.stepOut();
                break;
            case DebugConstants.CMD_STOP:
                debugger.clearDebugPoints();
                debugger.resume();
                break;
            case DebugConstants.CMD_SET_POINTS:
                // we expect { "command": "SET_POINTS", points: [{ "fileName": "sample.bal", "line" : 5 },{...},{...}]}
                debugSession.setBreakPoints(getPoints(json));
                sendAcknowledge(this.debugSession, "Debug points updated");
                break;
            case DebugConstants.CMD_START:
                if (this.waitingForClient) {
                    executionSem.release();
                    this.waitingForClient = false;
                    sendAcknowledge(this.debugSession, "Debug started.");
                }
                break;
            default:
                MessageDTO message = new MessageDTO();
                message.setCode(DebugConstants.CODE_INVALID);
                message.setMessage(DebugConstants.MSG_INVALID);
                debugServer.pushMessageToClient(debugSession, message);
        }
    }

    /**
     * Set debug channel.
     *
     * @param channel the channel
     */
    public void addDebugSession(Channel channel) {
        //@todo need to handle multiple connections
        this.debugSession = new DebugSession(channel);
        sendAcknowledge(this.debugSession, "Channel registered.");
    }

    private ArrayList<NodeLocation> getPoints(String json) {
        ArrayList<NodeLocation> list = new ArrayList<NodeLocation>();
        JsonParser parser = new JsonParser();
        JsonElement jsonTree = parser.parse(json);
        JsonObject command2 = jsonTree.getAsJsonObject();

        JsonArray points = command2.get("points").getAsJsonArray();
        for (int i = 0; i < points.size(); i++) {
            JsonObject tmp = points.get(i).getAsJsonObject();
            list.add(new NodeLocation(tmp.get("fileName").getAsString(), tmp.get("line").getAsInt()));
        }
        return list;
    }

    public void holdON() {
        try {
            while (!debugSession.getDebugger().isDone()) {
                    sleep(100);
            }
        } catch (InterruptedException e) {
            //@todo handle error
        }
    }

    public void setDebugger(BLangExecutionDebugger debugger) {
        // if we are handling multiple connections
        // we need to check and set to correct debug session
        if (null == this.debugSession) {
            throw new IllegalStateException("Debug session has not initialize, Unable to set debugger.");
        }

        this.debugSession.setDebugger(debugger);
    }

    public boolean isDebugSessionActive() {
        return (this.debugSession != null);
    }


    /**
     * Send a message to the debug client when a breakpoint is hit.
     *
     * @param debugSession
     * @param breakPointInfo
     */
    public void notifyDebugHit(DebugSession debugSession, BreakPointInfo breakPointInfo) {
        MessageDTO message = new MessageDTO();
        message.setCode(DebugConstants.CODE_HIT);
        message.setMessage(DebugConstants.MSG_HIT);
        message.setLocation(breakPointInfo.getHaltLocation());
        message.setFrames(breakPointInfo.getCurrentFrames());
        debugServer.pushMessageToClient(debugSession, message);
    }


    public void notifyComplete(DebugSession debugSession) {
        MessageDTO message = new MessageDTO();
        message.setCode(DebugConstants.CODE_COMPLETE);
        message.setMessage(DebugConstants.MSG_COMPLETE);
        debugServer.pushMessageToClient(debugSession, message);
    }

    public void notifyExit(DebugSession debugSession) {
        MessageDTO message = new MessageDTO();
        message.setCode(DebugConstants.CODE_EXIT);
        message.setMessage(DebugConstants.MSG_EXIT);
        debugServer.pushMessageToClient(debugSession, message);
    }

    public void sendAcknowledge(DebugSession debugSession, String messageText) {
        MessageDTO message = new MessageDTO();
        message.setCode(DebugConstants.CODE_ACK);
        message.setMessage(messageText);
        debugServer.pushMessageToClient(debugSession, message);
    }
}

File: tool-plugins/vscode/plugin/debugger/DebugManager/DebugManager.js
/**
 * Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

const EventEmitter = require('events');
const DebugChannel = require('./DebugChannel');
const DebugPoint = require('./DebugPoint');

/**
 * DebugManager
 * @class DebugManager
 * @extends {EventEmitter}
 */
class DebugManager extends EventEmitter {
    /**
     * Creates an instance of DebugManager.
     *
     * @memberof DebugManager
     */
    constructor() {
        super();
        this.debugPoints = [];
        this.channel = undefined;
        this.active = false;

        this.on('breakpoint-added', () => { this.publishBreakpoints(); });
        this.on('breakpoint-removed', () => { this.publishBreakpoints(); });

        this.on('execution-ended', () => {
            this.active = false;
        });
        this.on('session-error', () => {
            this.active = false;
        });
    }

    get active() {
        return this._active;
    }

    set active(active) {
        this._active = active;
    }

    /**
     * Send call to backend to step in
     *
     * @memberof DebugManager
     */
    stepIn() {
        const message = { command: 'STEP_IN', threadId: this.currentThreadId };
        this.channel.sendMessage(message);
    }
    /**
     * Send call to backend to step out
     *
     * @memberof DebugManager
     */
    stepOut() {
        const message = { command: 'STEP_OUT', threadId: this.currentThreadId };
        this.channel.sendMessage(message);
    }
    /**
     * Send call to backend to stop debugging
     *
     * @memberof DebugManager
     */
    stop() {
        this.emit('session-ended');
        this.channel.close();
    }
    /**
     * Send call to backend to step over
     *
     * @memberof DebugManager
     */
    stepOver() {
        const message = { command: 'STEP_OVER', threadId: this.currentThreadId };
        this.channel.sendMessage(message);
    }
    /**
     * Send call to backend to resume execution
     *
     * @memberof DebugManager
     */
    resume() {
        const message = { command: 'RESUME', threadId: this.currentThreadId };
        this.channel.sendMessage(message);
    }
    /**
     * Send call to backend to start debugging
     *
     * @memberof DebugManager
     */
    startDebug() {
        const message = { command: 'START' };
        this.channel.sendMessage(message);
        this.emit('debugging-started');
    }

    /**
     * Send call to backend to kill debug server
     *
     * @memberof DebugManager
     */
    kill() {
        const message = { command: 'KILL' };
        this.channel.sendMessage(message);
    }

    /**
     * @param {Object} message - Process message from backend
     *
     * @memberof DebugManager
     */
    processMesssage(message) {
        if (message.code === 'DEBUG_HIT') {
            this.emit('debug-hit', message);
            this.currentThreadId = message.threadId;
        }
        if (message.code === 'COMPLETE' || message.code === 'EXIT') {
            this.emit('execution-ended');
        }
    }

    /**
     * @param {String} url - Debugger backend url
     *
     * @memberof DebugManager
     */
    connect(url, callback) {
        this.emit('connecting');
        if (url !== undefined || url !== '') {
            this.channel = new DebugChannel(url);
            this.channel.connect();
        }
        this.channel.on('onmessage', (message) => {
            this.processMesssage(message);
        });
        this.channel.on('connected', () => {
            this.active = true;
            callback();
        });
        this.channel.on('session-ended', e => {
            this.emit('execution-ended', e);
        });
        this.channel.on('session-terminated', () => {
            this.emit('execution-ended');
        });
        this.channel.on('connection-closed', () => {
            this.emit('execution-ended');
        });
        this.channel.on('session-error', e => {
            this.emit('session-error', e);
        });
    }
    /**
     * Reconnect to backend
     * @memberof DebugManager
     */
    reConnect() {
        this.connect(this.endpoint);
    }

    /**
     * @param {Object} options - Debug manager configs
     *
     * @memberof DebugManager
     */
    init(endpoint) {
        this.endpoint = endpoint;
    }
    /**
     * Add new breakpoint
     * @param {Integer} lineNumber - Line number of the breakpoint
     * @param {String} fileName - File name of the breakpoint
     *
     * @memberof DebugManager
     */
    addBreakPoint(lineNumber, fileName, packagePath) {
        // log.debug('debug point added', lineNumber, fileName, packagePath);
        const point = new DebugPoint({ fileName, lineNumber, packagePath });
        this.debugPoints.push(point);
        this.emit('breakpoint-added', point);
    }
    /**
     * Remove breakpoint
     * @param {Integer} lineNumber - Line number of the breakpoint
     * @param {String} fileName - File name of the breakpoint
     *
     * @memberof DebugManager
     */
    removeBreakPoint(lineNumber, fileName, packagePath) {
        // log.debug('debug point removed', lineNumber, fileName);
        const point = new DebugPoint({ fileName, lineNumber, packagePath });
        // _.remove(this.debugPoints, item => item.fileName === point.fileName && item.lineNumber === point.lineNumber);
        this.emit('breakpoint-removed', point);
    }

    /**
     * Publishes current breakpoints to backend
     *
     * @memberof DebugManager
     */
    publishBreakpoints() {
        try {
            const message = { command: 'SET_POINTS', points: this.debugPoints };
            this.channel.sendMessage(message);
        } catch (e) {
            // @todo log
        }
    }
    /**
     * Check a given position for breakpoint or not
     * @param {Integer} lineNumber - Line number of the breakpoint
     * @param {String} fileName - File name of the breakpoint
     * @returns {Boolean} true if has a breakpoint in the position
     *
     * @memberof DebugManager
     */
    hasBreakPoint(lineNumber, fileName) {
        return !!this.debugPoints.find(point => point.lineNumber === lineNumber && point.fileName === fileName);
    }

    /**
     * Returns breakpoints for a given file
     * @param {String} fileName - Name of the ballerina file
     * @returns {[Integer]} line numbers of breakpoints
     *
     * @memberof DebugManager
     */
    getDebugPoints(fileName) {
        const breakpoints = this.debugPoints.filter(breakpoint => breakpoint.fileName === fileName);

        const breakpointsLineNumbers = breakpoints.map(breakpoint => breakpoint.lineNumber);
        return breakpointsLineNumbers;
    }
    /**
     * Removes all breakpoints for a ballerina file
     * @param {any} fileName - Name of the ballerina file
     *
     * @memberof DebugManager
     */
    removeAllBreakpoints(fileName) {
        this.debugPoints = this.debugPoints.filter(item => item.fileName !== fileName);
        this.publishBreakpoints();
    }

    /**
     * Create and return breakpoint object
     * @param {Integer} lineNumber - Line number of the breakpoint
     * @param {String} fileName - File name of the breakpoint
     * @returns {DebugPoint}
     *
     * @memberof DebugManager
     */
    createDebugPoint(lineNumber, fileName) {
        return new DebugPoint({ fileName, lineNumber });
    }
}

module.exports = DebugManager;
